### ISSUE TITLE
[red-knot] Silence *all* diagnostics in unreachable code?

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ruff#17388]
Our current approach to silencing only a subset of diagnostics in unreachable sections of code has some issues. They are [documented here](https://github.com/astral-sh/ruff/blob/main/crates/red_knot_python_semantic/resources/mdtest/unreachable.md#limitations-of-the-current-approach). In short, we still emit false positives in some cases:
```py
if False:
    x: int = 1
else:
    x: str = "a"

if False:
    # TODO We currently emit a false positive here:
    # error: [invalid-assignment] "Object of type `Literal["a"]` is not assignable to `int`"
    other: int = x
else:
    other: str = x
```

We should re-evaluate if it makes more sense to silence *all* diagnostics (possibly with a very limited set of exceptions, like `reveal_type`-messages) in unreachable sections.
